File: py/517.super-washing-machines.py
#
# @lc app=leetcode id=517 lang=python3
#
# [517] Super Washing Machines
#

# @lc code=start
class Solution:
    def findMinMoves(self, machines: List[int]) -> int:
        sum_val = sum(machines)
        width = len(machines)
        if sum_val % width:
            return -1
        required_val = sum_val//width
        res = 0
        left_sum = 0
        for i in range(width):
            pressure = left_sum-i*required_val
            left_sum += machines[i]
            if pressure < 0:
                # pass to left : -pressure
                # pass to right : max(machines[i] + pressure - required_val, 0)
                res = max(res, -pressure +
                          max(machines[i] + pressure - required_val, 0))
            else:
                # pass to right and receive : pressure
                res = max(res, pressure)
        return res


# @lc code=end

File: 517.super-washing-machines.java
import java.util.Arrays;

/*
 * @lc app=leetcode id=517 lang=java
 *
 * [517] Super Washing Machines
 */

// @lc code=start
class Solution {
    public int findMinMoves(int[] machines) {
        int sum_val = Arrays.stream(machines).sum();
        int width = machines.length;
        if (sum_val % width != 0) {
            return -1;
        }
        int required_val = sum_val / width;
        int res = 0;
        int left_sum = 0;
        for (int i = 0; i != width; ++i) {
            int pressure = left_sum - i * required_val;
            left_sum += machines[i];
            if (pressure < 0) {
                // pass to left : -pressure
                // pass to right : max(machines[i] + pressure - required_val, 0)
                res = Math.max(res, -pressure + Math.max(machines[i] + pressure - required_val, 0));
            } else {
                // pass to right and receive : pressure
                res = Math.max(res, pressure);
            }
        }
        return res;
    }
}
// @lc code=end
